tools/capture-test.py

```diff
@@ -0,0 +1,373 @@
+    alsa.snd_pcm_close(pcm)
+
+def pcm_play_chirp():
+    pcm = c.c_long(0)
+    dev = f"hw:{opts.card},{opts.pcm}".encode("ascii")
+    alsa.snd_pcm_open(c.byref(pcm), dev, alsa.PCM_STREAM_PLAYBACK, 0)
+    pcm_init_stream(pcm, opts.rate, opts.chan, alsa.PCM_FORMAT_S16_LE,
+                    alsa.PCM_ACCESS_MMAP_INTERLEAVED)
+
+    (chirp, chirp_frames) = gen_chirp_s16le(opts.rate, opts.chan)
+
+    # Reset the stream and queue up as much data as will fit in the
+    # ring buffer
+    area = alsa.pcm_channel_area_t()
+    offset = c.c_ulong()
+    frames = c.c_ulong(opts.rate)
+    ring_frames = 0
+    alsa.snd_pcm_prepare(pcm)
+    alsa.snd_pcm_reset(pcm)
+    while True:
+        alsa.snd_pcm_avail_update(pcm)
+        alsa.snd_pcm_mmap_begin(pcm, c.byref(area), c.byref(offset), c.byref(frames))
+        committed = alsa.snd_pcm_mmap_commit(pcm, offset, frames)
+        ring_frames += committed
+        if committed == 0:
+            break
+
+    silence = bytes(2 * opts.chan * ring_frames)
+
+    # Start up the stream, spin until there is space in the buffer,
+    # write the chirp.  This minimizes client-side overhead like
+    # stream startup.  Then immediately take a timestamp and write
+    # silence for one full cycle (to be 100% sure the buffer can't
+    # wrap and chirp twice).
+    alsa.snd_pcm_start(pcm)
+    while alsa.snd_pcm_avail(pcm) < chirp_frames:
+        pass
+    pre_buffered = ring_frames - alsa.snd_pcm_avail(pcm)
+    f = alsa.snd_pcm_mmap_writei(pcm, chirp, chirp_frames)
+    chirp_sent = time.perf_counter()
+
+    n = 0
+    while n < ring_frames:
+        n += alsa.snd_pcm_mmap_writei(pcm, silence, ring_frames)
+    alsa.snd_pcm_drain(pcm)
+    alsa.snd_pcm_close(pcm)
+
+    # Correct chirp_sent for buffered data!
+    chirp_sent += pre_buffered / opts.rate
+    return chirp_sent
+
+# Returns an array of tuples of (timestamp, bytes), no processing done
+# here for performance reasons, just one heap allocation and copy.
+def pcm_do_capture(duration):
+    pcm = c.c_long(0)
+    fmt = alsa.PCM_FORMAT_S32_LE if opts.capbits == 32 else alsa.PCM_FORMAT_S16_LE
+    capsz = 4 if opts.capbits == 32 else 2
+    dev = f"hw:{opts.card},{opts.cap}".encode("ascii")
+    alsa.snd_pcm_open(c.byref(pcm), dev, alsa.PCM_STREAM_CAPTURE, 0)
+    pcm_init_stream(pcm, opts.rate, opts.capchan, fmt, alsa.PCM_ACCESS_RW_INTERLEAVED)
+    frames_remaining = duration * opts.rate
+    buf_frames = int(opts.rate / 1000) # 1ms blocks
+    fsz = opts.capchan * capsz
+    buf = bytearray(fsz * buf_frames)
+    addr = c.c_ulong(c.addressof((c.c_byte * 1).from_buffer(buf)))
+    buflist = []
+    buf_frames = c.c_ulong(buf_frames)
+    while frames_remaining > 0:
+        f = alsa.snd_pcm_readi(pcm, addr, buf_frames)
+        t = time.perf_counter()
+        frames_remaining -= f
+        buflist.append((t, bytes(buf[0:f * fsz])))
+    return buflist
+
+# A programmatically-detectable chirp/pop signal for testing latency.
```

Comment:
not following which latency you are measuring, does this not add playback+capture continuous latencies together?

Reply:
Actually what I'm **really** measuring is a whitebox test inside AEC, which is doing essentially the same thing this python code is doing to watch the chirp go by.  This is here as a general smoke test from userspace.  But to be specific: it's measuring the time between the arrival of the chirp signal at the head of the userspace mmap buffer to the return from snd_pcm_readi() that contains it.

Reply:
Not following your explanation, sorry.

Reply:
1. Fill the mmap playback buffer[1] with silence
2. Start playing
3. Spin until there is just enough space in the buffer to accept the chirp, write it
4. Check buffer stats and take timestamp.

Now we know exactly how many samples are "ahead" of the chirp but still in host memory, so we can subtract them from the measured time it takes the chirp to appear on the capture stream.  You can't do this with snd_pcm_writei() as the buffering is internal to the userspace API.

[1] Which is obviously the actual DMA buffer, though the API doesn't promise this.

Reply:
not all DMAs provide a precise position, it's a recurring problem with audio - and it's only made worse with the bursty nature of the fimware-controlled DMAs.

You probably have at least one ms of error each way for the Intel HDaudio hw. no idea for others.

Reply:
No doubt.  But we do what we can.  The point was just to not have an unmeasurable 0-85ms of latency in the userspace process, or to have to guess at "how long after the first writei() call does the DMA stream start running", etc...  These numbers aren't the whole story, but they're at least reliable.

FWIW: I'm getting a pretty consistent 30-31ms round trip on MTL with AEC enabled, which seems not totally unreasonable (10ms buffer for AEC alone, plus 0-1ms getting into and out of the DP thread due to scheduling cadence mismatch) but also that it probably has room for improvement.    mt8195 is a disaster in comparison: 100ms on the device I'm using, which AFAICS isn't in the firmware at all but in some hardware component (you can also hear it squash short chirp bursts, which is why the default length is so large).

Reply:
I think this is ok with suitable disclaimers. E.g. the 30-31ms number sounds in the right ballpark and matches data got (without AEC) on with ALSA latency tester https://www.alsa-project.org/wiki/Test_latency.c and jack-delay https://wiki.archlinux.org/title/Professional_audio#Latency_verification . Both tools have larger more dependencies, but can be used on any Linux setup to verify the numbers.